### DISCUSSION TITLE
Windows startup time

### DISCUSSION THREAD

**Discussion by chichid** [wezterm/wezterm#2637] [General]
I noticed that on my windows machine, Wezterm takes more time to startup than on my mac. 
At this point I'm just wondering if that's to be expected or not.
When running in debug mode, the time between the first log and when Opengl is initialized for the first time is huge compared to mac (something like 440 Milliseconds). The time to first render is 656ms. 
Is that normal?

``
"13:26:31.119  DEBUG  config > Reloaded configuration! generation=1" 
``
...
``
13:26:31.559  DEBUG  wezterm_gui::termwindow                > OpenGL initialized! NVIDIA GeForce GTX 970/PCIe/SSE2 4.5.0 NVIDIA 516.94 wezterm version: 20220905-102802-7d4b8249
``
...
``
13:26:31.775  DEBUG  wezterm_gui::termwindow::render        > paint_impl before call_draw elapsed=551.2µs
``

**Comment by wez** [comment]
If you run `wezterm -n` (to skip loading any wezterm config), is it faster?

**Reply by chichid** [reply]
@wez nope, actually slightly slower because in my config I specify the font explicitly to be JetBrains Mono and disable the system font lookup.
I think it may be an Nvidia Driver problem. I have the same problem in glium and glutin where it takes almost 200ms just to initialize opengl.
It'll be a bit surprising though, the gtx 970 is still a decent GPU.

**Reply by SuperSandro2000** [reply]
``-n`` speeds it up by ~50 ms.

**Comment by chichid** [comment]
Can anyone with a windows machine (or ideally Nvidia GTX + Windows) run set WEZTERM_LOG=DEBUG && wezterm.exe to compare? I'll appreciate it a lot!

**Reply by SuperSandro2000** [reply]
Also takes ~400 ms for me.

**Reply by chichid** [reply]
ok, thanks for confirming, so I guess it's a windows/nvidia thing. Geez I'm reading here that the core problem is that nvidia loads a profile data base of around 3000KB to optimize itself for some games, which slows down its opengl initialization,  https://hero.handmade.network/forums/code-discussion/t/2503-%5Bday_235%5D_opengls_pixel_format_takes_a_long_time

**Comment by SuperSandro2000** [comment]
Why is the config reloaded three times at startup?

```
04:27:17.593  DEBUG  config > Reloaded configuration! generation=1
04:27:17.723  DEBUG  config > Reloaded configuration! generation=1
04:27:17.755  DEBUG  config > Reloaded configuration! generation=2
04:27:17.757  DEBUG  wezterm_client::discovery::windows > published gui path as gui-sock-29860
04:27:17.766  DEBUG  wezterm_gui::frontend              > workspace is default, fixup windows
```

Also here are 120ms being used in ``SetDecPrivateMode(Code(FocusTracking)))``

```
04:27:17.787  DEBUG  wezterm_font::ftwrap               > set_char_size computing 10 dpi=96 (pixel height=13.333333333333334)
04:27:17.795  DEBUG  wezterm_term::terminalstate::performer > perform CSI(Mode(SetDecPrivateMode(Code(Win32InputMode))))
04:27:17.795  DEBUG  wezterm_term::terminalstate::performer > perform CSI(Mode(SetDecPrivateMode(Code(FocusTracking))))
04:27:17.919  DEBUG  wezterm_term::terminalstate::performer > perform CSI(Mode(ResetDecPrivateMode(Code(ShowCursor))))
04:27:17.919  DEBUG  wezterm_term::terminalstate::performer > perform CSI(Edit(EraseInDisplay(EraseDisplay)))
04:27:17.919  DEBUG  wezterm_term::terminalstate::performer > perform CSI(Sgr(Reset))
04:27:17.919  DEBUG  wezterm_term::terminalstate            > Reset
04:27:17.919  DEBUG  wezterm_term::terminalstate::performer > perform CSI(Sgr(Foreground(PaletteIndex(1))))
04:27:17.919  DEBUG  wezterm_term::terminalstate            > Foreground(PaletteIndex(1))
```

**Reply by wez** [reply]
Config can be evaluated several times due to a number of factors.

The log lines aren't tracing spans: the time elapsed between two log lines doesn't mean that the thing on the first log line took up 100% of that time delta.  For the FocusTracking/ShowCursor lines, it suggests to me that there was a 200ms interval between the application sending those escape sequences, not that it took wezterm 200ms to process them.

**Reply by SuperSandro2000** [reply]
Good to know. What else could we do to improve this?

**Reply by chichid** [reply]
for the GPU issue, I'm facing, I think it's an Nvidia issue. But no one cares, because it's meant for gaming. Where 400ms initial loading doesn't seem to be a big deal. 
I believe the only solution would be to support different GUI backends such as "direct2d" on Windows. Pretty much an overkill and not a deal breaker for me (even though it would have been really awesome if wezterm could load as fast as the native windows prompt)
@wez for the rest same as @SuperSandro2000 I think it'll be good to defer some of those initializations until the first window is shown (maybe even showing the window with it's background and then initializing/connecting opengl, if that's even possible).
If you're interested into improving this area I could start a side project within a fork to explore that.

**Reply by wez** [reply]
I think the first thing to do is to understand what is slow before deciding how to address the slowness. I don't have time to look into this, and with my usage (opening a window once at startup, then keeping it open forever), this is never a practical issue for me.

If you'd like to put in some time analyzing where the time is spent and reporting your findings, I'd be happy to discuss them with you and suggest next steps.

**Reply by chichid** [reply]
@wez what's a good way to measure, if the logs are not the way to go?

**Reply by wez** [reply]
What I usually do is bracket things with something like:

```rust
let start = Instant::now();
// thing happens here
log::trace!("thing took {:?}", start.elapsed());
```

to help surface what's taking the time.  A lot of those end up not being worth keeping, but you'll see a few places in the code that do that today.

**Reply by chichid** [reply]
ok sounds good, this is now my side project!

**Reply by chichid** [reply]
@wez I think I have found the problem, there is a huge amount of time taken in opengl initialization under the enable_opengl function within the windows module in the crate "window". Something like 300-400ms.
I figured while tracing that if I set prefer_egl=true in my config.lua, it'll use Angle, which I did and boom, now the opengl initialization is in the order of 50-100ms. 
Is this a bug, because I thought prefer_egl defaults to true?

**Reply by wez** [reply]
That's a good find!  So it sounds like the opengl drivers on your system are slow to initialize.
What are they? (what does `CTRL-SHIFT-L` show about them?)

The code defaults `prefer_egl = !cfg(windows)`, so the docs are a little wonky wrt. what actually happens now.
571c137955f63138c46bca3b4c3d31a68cf8ca08 changed the default for windows.

**Reply by chichid** [reply]
@wez it's the Nvidia 970 GTX official drivers.

My wild guess is that startup time, for Nvidia is secondary since they focus more on the runtime performance. Maybe even worse, that they focus less on Opengl now. I have tried different samples with DirectX and Opengl (Glium vs the directX samples from microsoft), and noticed that pure DirectX loads instantly, while Opengl always take at best 100ms to load.

Forcing prefer_egl to true resolves the slightly visible delay I started noticing for whatever reason.

**Reply by chichid** [reply]
![image](https://user-images.githubusercontent.com/6054451/199277946-ac2147a5-3c1c-4e6a-bca8-95257655d2d5.png)
